### PR TITLE
twister: Fix tests using required_snippets

### LINKED ISSUE #63011: Snippets as extra arguments on Twister CLI leads to early CMake error
**Describe the bug**

Whenever you call Twister with the command line argument `--extra-args SNIPPET="<valid-snippet-name>"`, CMake aborts with the error message `zephyr_set(... SCOPE snippets) doesn't exist.`. This happens very early on in Zephyr bootstrapping, more specifically when including `cmake/modules/snippets.cmake`.

The analogous usage for a simple build, on the other hand, does not result in any error message, so both `west build -S <valid-snippet-name>` and `west build -- -DSNIPPET="<valid-snippet-name>"` work as described and documented. So far, I myself was not able to find the cause of the non-existent scope for snippets in the case of Twister and need help from experts.

**To Reproduce**

Here is a more fancy console log:

```
west twister --verbose --inline-logs --cmake-only \
     --extra-args SNIPPET="cdc-acm-console" \
     --platform nucleo_f767zi \
     --test tests/drivers/uart/uart_async_api/drivers.uart.async_api
```

…breaks with (CMake output in `build.log`):

```
-- Application: /.../workspace-zephyr/zephyr/tests/drivers/uart/uart_async_api
-- CMake version: 3.27.4
-- Found Python3: /.../workspace-zephyr/.env/bin/python (found suitable version "3.8.10", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /home/user/.cache/zephyr
-- Zephyr version: 3.4.99 (/.../workspace-zephyr/zephyr)
-- Found west (found suitable version "1.1.0", minimum required is "0.14.0")
-- Board: nucleo_f767zi
CMake Error at zephyr/cmake/modules/extensions.cmake:2764 (message):
  zephyr_set(...  SCOPE snippets) doesn't exists.
Call Stack (most recent call first):
  twister-out/nucleo_f767zi/tests/drivers/uart/uart_async_api/drivers.uart.async_api/zephyr/snippets_generated.cmake:25 (zephyr_set)
  zephyr/cmake/modules/snippets.cmake:94 (include)
  zephyr/cmake/modules/snippets.cmake:121 (zephyr_process_snippets)
  zephyr/cmake/modules/zephyr_default.cmake:129 (include)
  zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:75 (include)
  zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  zephyr/cmake/package_helper.cmake:115 (find_package)
```

**Expected behavior**

The expectation here would be that Twister behaves the same way with snippets as it does when passing a shield stack via extra arguments. Both `--extra-args SNIPPET="<valid-snippet-name>"` and `--extra-args SHIELD="<valid-shield-name>"` should not lead to any CMake errors, alone or together.

**Impact**

Announcement for the community, but showstopper for [Bridle](https://github.com/tiacsys/bridle). Currently I'm trying to convert a [board-specific shield with CPU clock tweaks](https://github.com/tiacsys/bridle/tree/main/boards/shields/can_timing_adj) to a snippet so that the default [Zephyr CAN timing test](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/can/timing) passes.

**Logs and console output**
See above.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Zephyr SDK 0.16.1
 - main branch, commit SHA 341590545ac80b7a2abd9902b7f10d577323ccde

**Additional context**
Non.

### PR DESCRIPTION
**Pull request by nordicjm**
Fixes an issue whereby twister would run tests with a
    required_snippet set and it would apply the snippet during the
    Kconfig filtering stage without all the modules being available

    Adds support for using the Zephyr module snippet_root variable
    in twister for required_snippets

Fixes #63011

### CODE REVIEW

**Comment by rexut** [comment]
> Adds support for twister automatically checking which modules have snippet directories and including them when searching for required snippets
> 
> Fixes #63011

I agree, it'll fix the test plan strategie to find snippets also in Zephyr modules.

**REVIEW: +1**

**Comment by nordicjm** [comment]
@rexut I've updated it, can you give it another try?

**Comment by rexut** [comment]
> @rexut I've updated it, can you give it another try?

I do so, but no success. The special `SCOPE snippets` still doesn't exists. I'm also out of focus where in all the cmak modules stuff this scope will be created in an always proper way, if `west build …` will be used instead of `west twister …`.

**Comment by nordicjm** [comment]
Maybe that is an issue with the alternative root, as I can't reproduce it just using twister with a modified hello_world. @gmarull any clue on this issue (from https://github.com/zephyrproject-rtos/zephyr/pull/57806) ?

**Comment by nordicjm** [comment]
@rexut I think it should now finally be fixed!

File: cmake/modules/snippets.cmake [hunk 1]
```diff
@@ -65,6 +65,15 @@ function(zephyr_process_snippets)
...
     list(APPEND real_snippet_root ${real_snippet_dir})
   endforeach()
+
+  # Search through all modules for any that have a snippets directory
+  foreach(module_name ${ZEPHYR_MODULE_NAMES})
+    string(TOUPPER "${module_name}" module_name)
+    if(EXISTS "${ZEPHYR_${module_name}_MODULE_DIR}/snippets")
+      list(APPEND real_snippet_root "${ZEPHYR_${module_name}_MODULE_DIR}")
+    endif()
+  endforeach()
```
Comment (tejlmand) [inline cmake/modules/snippets.cmake:75]: we should not automatically loop all modules for a snippet root.

The `zephyr/module.yml` file of the module should instead use:
```
build:
  settings:
    snippet_root: <relative-path-to-root>
```
Ref: https://docs.zephyrproject.org/latest/develop/modules.html#build-settings
> snippet_root: Contains additional snippets that are available for use. These snippets must be defined in snippet.yml files underneath the <snippet_root>/snippets folder. 

```suggestion
```

File: doc/releases/release-notes-3.5.rst [hunk 2]
```diff
@@ -128,6 +128,9 @@ Build system and infrastructure
   be used to include a snippet when a test is ran (and exclude any boards from
   running that the snippet cannot be applied to).
 
+* Added support for snippet roots being automatically detected and used if they
+  exist in module directories.
+
```
Comment (tejlmand) [inline doc/releases/release-notes-3.5.rst:133]: This is already supported through the use of `zephyr/module.yml`.

```suggestion
```

File: scripts/pylib/twister/twisterlib/testplan.py [hunk 3]
```diff
@@ -309,6 +310,11 @@ def handle_modules(self):
         modules_meta = parse_modules(ZEPHYR_BASE)
         self.modules = [module.meta.get('name') for module in modules_meta]
 
+        # Check each module for a snippets directory, then add to a list if it is found
+        for module in modules_meta:
+            snippet_path = module.project + '/snippets'
+            if os.path.exists(snippet_path):
+                self.snippet_roots.append(Path(module.project))
```
Comment (tejlmand) [inline scripts/pylib/twister/twisterlib/testplan.py:317]: We should not have a deviating implementation in twister where snippets are picked up differently than in the build system.

The build system uses the information provided in `zephyr/module.yml` but this code checks for the existence of a folder regardless of the setting in `zephyr/module.yml`.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Please look at the current support for snippet_root in modules:
https://docs.zephyrproject.org/latest/develop/modules.html#build-settings

https://github.com/zephyrproject-rtos/zephyr/blob/72ec6ce4db682c89bdcdd5688294ca075b8832a0/scripts/zephyr_module.py#L275-L276

**Review by tejlmand — DISMISSED** [review]

**Pull request closed without merging**